### PR TITLE
Use composable nodes and add segway controller

### PR DESCRIPTION
**Pull request by bmchalenv**

### CODE REVIEW

File: ROS2/src/segwayrmp/CMakeLists.txt [hunk 1]
```diff
@@ -55,39 +45,36 @@ else()
...
-
-add_executable(drive_segway_sample
-  tools/drive_segway_sample.cpp
-)
-target_include_directories(drive_segway_sample PUBLIC
-  $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
-  $<INSTALL_INTERFACE:include>
+rclcpp_components_register_node(segway_chassis_component PLUGIN "robot::Chassis" EXECUTABLE SmartCar)
+
+# add_executable(drive_segway_sample
```
Comment (sgillen) [inline ROS2/src/segwayrmp/CMakeLists.txt:62]: Did you mean to comment these out? Looks to me like you meant to delete them.

File: ROS2/src/segwayrmp/include/segwayrmp/robot.h [hunk 2]
```diff
@@ -54,17 +58,25 @@ namespace robot
...
+            std::string odom_frame_name_;
+            bool publish_tf_;
 
             std::string bins_directory;
             std::string central_version;
             std::string motor_version;
 
             tf2::Quaternion odom_quat;
-            std::shared_ptr<tf2_ros::StaticTransformBroadcaster> odom_broadcaster;
+            std::shared_ptr<tf2_ros::TransformBroadcaster> odom_broadcaster;
```
Comment (sgillen) [inline ROS2/src/segwayrmp/include/segwayrmp/robot.h:79]: What's the reason for this change exactly?

File: ROS2/src/segwayrmp/src/robot.cpp [hunk 3]
```diff
@@ -39,6 +41,57 @@ uint8_t OdomEulerXy_update;
 uint8_t OdomEulerZ_update;
 uint8_t OdomVelLineXy_update;
 
+static std::map<uint32_t, std::string> hostErrors = {
```
Comment (sgillen) [inline ROS2/src/segwayrmp/src/robot.cpp]: Will the compiler let you make these constexpr? If not then at least const?

File: ROS2/src/segwayrmp/src/robot.cpp [hunk 4]
```diff
@@ -72,8 +125,26 @@ void EventPubData(int event_no)
     robot::Chassis::pub_event_callback(event_no);
 }
 
+std::string convert_to_hex_str(uint32_t value) {
+    std::stringstream ss;
+    ss << "0x" << std::setw(8) << std::setfill('0') << std::hex << value;
+    return ss.str();
+}
+
+std::string get_error_info(std::map<uint32_t, std::string> error_map, uint32_t key) {
```
Comment (sgillen) [inline ROS2/src/segwayrmp/src/robot.cpp]: These should be const references

File: ROS2/src/segwayrmp/src/robot.cpp [hunk 5]
```diff
@@ -425,6 +531,54 @@ void Chassis::timer_1hz_callback(void)
...
+    diagnostic_status.values.push_back(bms_error_kv);
+    if (error_code_fb.host_error || error_code_fb.central_error || error_code_fb.left_motor_error
+        || error_code_fb.right_motor_error || error_code_fb.bms_error) {
+        diagnostic_status.level = diagnostic_msgs::msg::DiagnosticStatus::ERROR;
+    } else {
+        diagnostic_status.level = diagnostic_msgs::msg::DiagnosticStatus::OK;
+    }
+    diagnostic_status.name = "segway_rmp";
+    diagnostics_msg.status.push_back(diagnostic_status);
+    diagnostics_pub->publish(diagnostics_msg);
```
Comment (sgillen) [inline ROS2/src/segwayrmp/src/robot.cpp:581]: This is awesome 💪

File: ROS2/src/segwayrmp/include/segwayrmp/robot.h (hunk 2 above)
Comment (bmchalenv) [reply ROS2/src/segwayrmp/include/segwayrmp/robot.h:79]: Odom is not  a static TF, it publishes at 40hz. The static transform broadcaster assumes the TF is only broadcast once and does not use time information/simplifies the TF.

File: ROS2/src/segwayrmp/CMakeLists.txt (hunk 1 above)
Comment (bmchalenv) [reply ROS2/src/segwayrmp/CMakeLists.txt:62]: I commented it out right now since it wouldn't compile. It did not compile before I modified anything either. I didn't want to delete it because segway may still be using that exe.

File: ROS2/src/segwayrmp/src/robot.cpp (hunk 3 above)
Comment (bmchalenv) [reply ROS2/src/segwayrmp/src/robot.cpp]: Made it const for now, constexpr isn't possible with an std::map (it errors).

File: ROS2/src/segwayrmp/src/robot.cpp [hunk 6]
```diff
@@ -41,13 +41,13 @@ uint8_t OdomEulerXy_update;
 uint8_t OdomEulerZ_update;
 uint8_t OdomVelLineXy_update;
 
-static std::map<uint32_t, std::string> hostErrors = {
+const std::map<uint32_t, std::string> hostErrors = {
```
Comment (sgillen) [inline ROS2/src/segwayrmp/src/robot.cpp]: Not that it really matters but you could make this both static and const.

Comment (sgillen) [inline ROS2/src/segwayrmp/src/robot.cpp]: I also really meant for the string to be const in this case. Keys to a map are implicitly const so no need there.
